### PR TITLE
util: add datum type

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @coocood @qiuyesuifeng PTAL

### CODE REVIEW

**Comment by coocood** [comment]
LGTM

**Comment by ngaut** [comment]
@shenli PTAL

**Comment by BusyJay** [comment]
@ngaut @qiuyesuifeng PTAL

**Comment by ngaut** [comment]
LGTM

File: src/util/codec/datum.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+pub fn decode_datum(buf: &[u8]) -> Result<(Datum, usize)> {
+    // TODO use read instead.
+    try!(codec::check_bound(buf, 1));
+
+    let datum;
+    let mut readed = 1;
+    match buf[0] {
+        INT_FLAG => {
+            let v = try!(number::decode_i64(&buf[1..]));
+            readed += 8;
```
Comment (qiuyesuifeng) [inline src/util/codec/datum.rs]: It is better to move to line 53.

File: src/util/codec/datum.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            datum = Datum::Null;
+        }
+        _ => unimplemented!(),
+    }
+    Ok((datum, readed))
+}
+
+/// `decode` decodes all datum from a byte slice generated by tidb.
+pub fn decode(buf: &[u8]) -> Result<(Vec<Datum>, usize)> {
+    try!(codec::check_bound(buf, 1));
```
Comment (qiuyesuifeng) [inline src/util/codec/datum.rs]: Seems needless to check here?

File: src/util/codec/datum.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                  _ => unimplemented!(),
+              }
+          })
+          .sum()
+}
+
+fn encode_bytes(buf: &mut [u8], value: &[u8], comparable: bool) -> Result<usize> {
+    let len = if comparable {
+        buf[0] = BYTES_FLAG;
+        try!(bytes::encode_bytes_to_buf(&mut buf[1..], value))
```
Comment (qiuyesuifeng) [inline src/util/codec/datum.rs]: Should we check bound here?

File: src/util/codec/datum.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    #[test]
+    fn test_datum_codec() {
+        let mut table = vec![
+			vec![Datum::I64(1)],
+			vec![Datum::U64(1), Datum::Bytes(b"123".to_vec()), Datum::I64(-1)],
+			vec![Datum::Null],
+		];
+
+        for vs in table.drain(..) {
+
```
Comment (qiuyesuifeng) [inline src/util/codec/datum.rs]: Remove empty line?

File: src/util/codec/datum.rs (hunk 2 above)
Comment (BusyJay) [inline src/util/codec/datum.rs]: It might be empty.

Comment (qiuyesuifeng) [inline src/util/codec/datum.rs]: Yes, i know, but i find that decode_datum has checked it also.

**Comment by BusyJay** [comment]
@qiuyesuifeng PTAL

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**